File: ObjC_Example/Native/NativeAdCarouselViewController.m
//
//  NativeAdCarouselViewController.m
//  ObjC_Example
//
//  Copyright © 2016年 FAN Communications. All rights reserved.
//

#import "NativeAdCarouselViewController.h"
#import "NativeAdCarouselCell.h"

static const int adRow = 3;
static const float cellPortrait = 325.f;
static const float cellLandscape = 200.f;

@interface NativeAdCarouselViewController ()<UITableViewDataSource, UITableViewDelegate>

@property (nonatomic) NSMutableArray<NSString *> *items;
@property (nonatomic) NSNumber *direction; // 端末向き

@end

@implementation NativeAdCarouselViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    
    self.items = [NSMutableArray array];
    for (int i = 0; i < 100; i++) {
        [self.items addObject:[NSString stringWithFormat:@"item%d", i + 1]];
    }
    
    // 画面向きの判定
    BOOL isLandscape;
    
    if (@available(iOS 13.0, *)) {
        UIWindowScene *windowScene = [[[UIApplication sharedApplication] windows] firstObject].windowScene;
        if (windowScene == nil) {
            isLandscape = NO;
        } else {
            isLandscape = UIInterfaceOrientationIsLandscape(windowScene.interfaceOrientation);
        }
    } else {
        isLandscape = UIInterfaceOrientationIsLandscape(UIApplication.sharedApplication.statusBarOrientation);
    }
    
    if (isLandscape == YES) {
        self.direction = @2;
    } else {
        self.direction = @1;
    }
}

- (void)didReceiveMemoryWarning {
    [super didReceiveMemoryWarning];
    // Dispose of any resources that can be recreated.
}

// 画面向きの判定
- (void)viewWillTransitionToSize:(CGSize)size withTransitionCoordinator:(id<UIViewControllerTransitionCoordinator>)coordinator
{
    [coordinator animateAlongsideTransition:^(id<UIViewControllerTransitionCoordinatorContext> context) {
        
        if (size.width <= size.height) {
            self.direction = @1;
        } else {
            self.direction = @2;
        }
        
        [self.tableView reloadData];
    } completion:^(id<UIViewControllerTransitionCoordinatorContext> context){
        [[NSNotificationCenter defaultCenter] postNotificationName:@"layoutUpdate" object:self.direction];
    }];
}

#pragma mark - Table view data source
- (NSInteger)numberOfSectionsInTableView:(UITableView *)tableView
{
    // Return the number of sections.
    return 1;
}

- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section
{
    // Return the number of rows in the section.
    return self.items.count;
}

- (CGFloat)tableView:(UITableView *)tableView heightForRowAtIndexPath:(NSIndexPath *)indexPath
{
    if (indexPath.row == adRow) {
        if ((self.direction).integerValue == 1) {
            return cellPortrait;
        } else {
            return cellLandscape;
        }
    } else {
        return 44.f;
    }
}

- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath
{
    if (indexPath.row == adRow) {
        static NSString *AdCellIdentifier = @"adcell";
        NativeAdCarouselCell *cell= [tableView dequeueReusableCellWithIdentifier:AdCellIdentifier];
        if (cell == nil) {
            cell = [[NativeAdCarouselCell alloc]initWithStyle:UITableViewCellStyleDefault reuseIdentifier:AdCellIdentifier];
            cell.direction = self.direction;
            cell.selectionStyle = UITableViewCellSelectionStyleNone;
            [cell initAd];
        }
        
        return cell;
    } else {
        static NSString *CellIdentifier = @"cell";
        UITableViewCell *cell = [tableView dequeueReusableCellWithIdentifier:CellIdentifier];
        if (cell == nil) {
            cell = [[UITableViewCell alloc]initWithStyle:UITableViewCellStyleDefault reuseIdentifier:CellIdentifier];
            cell.accessoryType = UITableViewCellAccessoryDisclosureIndicator;
        }
        if (@available(iOS 14.0, *)) {
            UIListContentConfiguration *content = [cell defaultContentConfiguration];
            [content setText:self.items[indexPath.row]];
            [cell setContentConfiguration:content];
        } else {
            cell.textLabel.text = self.items[indexPath.row];
        }
        
        return cell;
    }
}

#pragma mark - Table view delegate

- (void)tableView:(UITableView *)tableView didSelectRowAtIndexPath:(NSIndexPath *)indexPath
{
    //
}

@end

File: Sample/NendSDK_Swift_Sample/NativeAdCarouselViewController.swift
//
//  NativeAdCarouselViewController.swift
//  NendSDK_Sample
//
//  Copyright © 2016年 F@N Communications. All rights reserved.
//

import UIKit

let adPortraitWidth: CGFloat = 320 // 縦向き　広告横幅
let adPortraitHeight: CGFloat = 325 // 縦向き 広告高さ
let adLandscapeWidth: CGFloat = 580 // 横向き　広告横幅
let adLandscapeHeight: CGFloat = 200 // 横向き　広告高さ

class NativeAdCarouselViewController: UITableViewController {
    private let adRow : Int = 3
    private var items = [String]()
    private var direction : Int = 0
    
    deinit {
        print("NativeAdCarouselViewController: deinit")
    }
    
    override func viewDidLoad() {
        super.viewDidLoad()
        
        self.title = "Carousel"
        
        for i in 1...100 {
            self.items.append("Item\(i)")
        }
        
        let orientation = UIApplication.sharedApplication().statusBarOrientation
        switch (orientation) {
            case UIInterfaceOrientation.Portrait, UIInterfaceOrientation.PortraitUpsideDown:
                self.direction = 1
                break
            case UIInterfaceOrientation.LandscapeLeft, UIInterfaceOrientation.LandscapeRight:
                self.direction = 2
                break
            default:
                self.direction = 0
                break
        }
    }
    
    override func viewWillTransitionToSize(size: CGSize, withTransitionCoordinator coordinator: UIViewControllerTransitionCoordinator) {
        coordinator.animateAlongsideTransition({(UIViewControllerTransitionCoordinatorContext) in
                if size.width <= size.height {
                    self.direction = 1
                } else {
                    self.direction = 2
                }
                self.tableView.reloadData()
            },
            completion: {(UIViewControllerTransitionCoordinatorContext) in
                NSNotificationCenter.defaultCenter().postNotificationName("layoutUpdate", object: self.direction)
            }
        )
    }
    
    override func didReceiveMemoryWarning() {
        super.didReceiveMemoryWarning()
        // Dispose of any resources that can be recreated.
    }
    
    // MARK: - Table view data source
    override func numberOfSectionsInTableView(tableView: UITableView) -> Int {
        return 1
    }
    
    override func tableView(tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        return self.items.count
    }
    
    override func tableView(tableView: UITableView, heightForRowAtIndexPath indexPath: NSIndexPath) -> CGFloat {
        if indexPath.row == self.adRow {
            if self.direction == 1 {
                return adPortraitHeight
            } else {
                return adLandscapeHeight
            }
        }
        
        return 44.0
    }
    
    override func tableView(tableView: UITableView, cellForRowAtIndexPath indexPath: NSIndexPath) -> UITableViewCell {
        
        if indexPath.row == self.adRow {
            let rowId = "adcell"
            var cell = tableView.dequeueReusableCellWithIdentifier(rowId) as! NativeAdCarouselCell?
            if cell == nil {
                cell = NativeAdCarouselCell(style: UITableViewCellStyle.Default, reuseIdentifier: rowId)
                cell!.direction = self.direction
                cell!.initAd()
            }
            
            return cell!
        } else {
            let rowId = "cell"
            var cell = tableView.dequeueReusableCellWithIdentifier(rowId) as UITableViewCell?
            if cell == nil {
                cell = UITableViewCell(style: UITableViewCellStyle.Default, reuseIdentifier: rowId)
            }
            
            cell!.textLabel!.text = self.items[indexPath.row]
            return cell!
        }
    }
    
    // MARK: - Table view delegate
    override func tableView(tableView: UITableView, didSelectRowAtIndexPath indexPath: NSIndexPath) {
        //
    }

}